utils/networking/src/network_address.rs

```diff
@@ -0,0 +1,302 @@
+use thiserror::Error;
+
+use crate::{
+    ip_or_socket_address::IpOrSocketAddress,
+    resolvable_name::{NameResolutionError, ResolvableName},
+};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DomainName(String);
+
+impl Display for DomainName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl FromStr for DomainName {
+    type Err = DomainNameParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Note: the 'addr' crate also has the 'parse_dns_name' function, which considers
+        // '[1:2:3:4:5:6:7:8]' to be a domain name.
+        match addr::parse_domain_name(s) {
+            Ok(_) => Ok(DomainName(s.to_owned())),
+            Err(_) => Err(DomainNameParseError {
+                erroneous_domain_name: s.to_owned(),
+            }),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("Not a domain name: '{erroneous_domain_name}'")]
+pub struct DomainNameParseError {
+    erroneous_domain_name: String,
+}
+
+impl ResolvableName for DomainName {
+    type ResolvedAddress = IpAddr;
+
+    async fn resolve(&self) -> Result<impl Iterator<Item = IpAddr> + '_, NameResolutionError> {
+        // Note: for historical reasons, std::net::ToSocketAddrs, which is used internally
+        // by tokio::net::lookup_host, expects a pair of (host, port) and resolves it to
+        // SocketAddr, despite the fact that the port number is useless during name resolution.
+        // So, we specify the fake port number 0 here and then ignore it in the result.
+        Ok(tokio::net::lookup_host((self.0.as_str(), 0))
+            .await
+            .map_err(|error| NameResolutionError::CannotResolve {
+                resolvable_name: self.0.clone(),
+                error_str: error.to_string(),
+            })?
+            .map(|socket_addr| socket_addr.ip()))
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum NetworkAddress {
+    Ip(IpAddr),
+    Domain(DomainName),
+}
+
+impl NetworkAddress {
+    pub fn as_ip_address(&self) -> Option<&IpAddr> {
+        match self {
+            Self::Ip(addr) => Some(addr),
+            Self::Domain(_) => None,
+        }
+    }
+}
+
+impl ResolvableName for NetworkAddress {
+    type ResolvedAddress = IpAddr;
+
+    async fn resolve(
+        &self,
+    ) -> Result<impl Iterator<Item = Self::ResolvedAddress> + '_, NameResolutionError> {
+        match self {
+            Self::Ip(ip) => Ok(Either::Left(std::iter::once(*ip))),
+            Self::Domain(domain_name) => Ok(Either::Right(domain_name.resolve().await?)),
+        }
+    }
+}
+
+impl FromStr for NetworkAddress {
+    type Err = NetworkAddressParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(ip) = s.parse::<IpAddr>() {
+            Ok(NetworkAddress::Ip(ip))
+        } else {
+            Ok(NetworkAddress::Domain(s.parse::<DomainName>()?))
+        }
```

Comment:
I think we should do a little better than this... but if it's too much trouble, I'm OK with what we have. How hard is it to do a simple character check? We should only allow characters for FQDNs:

Letters (a-z, A-Z)
Digits (0-9)
Hyphens (-) - However, hyphens are not allowed as the first or last character of a label (part between the dots), nor can they appear consecutively.
Periods (.) - Periods are used to separate the labels in the domain name.

From the RFC: "The labels must follow the rules for ARPANET host names.  They must start with a letter, end with a letter or digit, and have as interior characters only letters, digits, and hyphen.  There are also some restrictions on the length.  Labels must be 63 characters or less.".

https://www.rfc-editor.org/rfc/rfc1035.txt

Reply:
Here we delegate to `impl FromStr for DomainName`; there, `addr::parse_domain_name` is used to parse the name, which should check its correctness.
I have a couple of checks for incorrect names in `network_address_printing_parsing`:
```
// Some garbage.
let err = "!@#$%^".parse::<NetworkAddress>().unwrap_err();
assert_eq!(
    err,
    NetworkAddressParseError::NotIpAddressOrDomainName("!@#$%^".to_owned())
);

// "Ipv6 address in brackets" (without a port) is not a valid address or domain name.
let err = "[1:2:3:4:5:6:7:8]".parse::<NetworkAddress>().unwrap_err();
assert_eq!(
    err,
    NetworkAddressParseError::NotIpAddressOrDomainName("[1:2:3:4:5:6:7:8]".to_owned())
);
``` 
I can add more if you think we need them.

Reply:
I've left it as is.